---
title: "OpenGWAS"
author: "Brian M. Schilder"
date: "`r Sys.Date()`"
csl: nature.csl
output:
  BiocStyle::html_document:
vignette: >
    %\VignetteIndexEntry{OpenGWAS} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---

# Install packages

To install the dev branch of `MAGMA.Celltyping`, use:

```{r, eval=FALSE, include=TRUE}
if(!require("remotes")) install.packages("remotes")
if(!require("here")) install.packages("here")
if(!require("BiocManager")) install.packages("BiocManager")


if(!require("MungeSumstats")) {
  BiocManager::install("MungeSumstats", dependencies = TRUE)
}
if(!require("EWCE") || packageVersion("EWCE")<1.3.1){
  remotes::install_github("NathanSkene/EWCE@bschilder_dev")
} 
if(!require("MAGMA.Celltyping")) {
  remotes::install_github("neurogenomics/MAGMA_Celltyping@bschilder_dev")
}
```


```{r setup, include=TRUE}
library(MAGMA.Celltyping)
library(MungeSumstats)

library(ggplot2)
library(dplyr)

source(here::here("code/utils.R"))

save_dir <- here::here("data/GWAS_sumstats")
dir.create(save_dir, showWarnings = FALSE, recursive = TRUE)
```

# Munge sum stats

## Find sum stats

```{r} 
metagwas <- MungeSumstats::find_sumstats(traits = c("Alzheimer","Parkinson",
                                                    "amyotrophic","dementia"))
knitr::kable(metagwas)
```

## Import sum stats

`MungeSumstats::import_sumstats` already supports parallelisation 
across multiple GWAS (by setting `nThread` >1). 
However, the log files would all 
overwrite one another. 

So we instead parellelise this step ourselves,
saving each summary stats file in it's own folder with a 
subdirectory for its respective log files.

```{r}
gwas_paths <- parallel::mclapply(metagwas$id, function(id){
  id_dir <- file.path(save_dir,id)
  dir.create(id_dir, showWarnings = FALSE, recursive = TRUE)
  log_dir <- file.path(id_dir,"logs")
  dir.create(log_dir, showWarnings = FALSE, recursive = TRUE)
  gwas <- MungeSumstats::import_sumstats(ids = id, 
                                       save_dir = id_dir, 
                                       nThread = 1, 
                                       parallel_across_ids = FALSE,
                                       force_new_vcf = FALSE,
                                       force_new = FALSE,
                                       #### Record logs
                                       log_folder_ind = TRUE,
                                       log_mungesumstats_msgs = TRUE,
                                       log_folder = log_dir)
  return(setNames(gwas, id))
}, mc.cores = 32) 
 
```


# Create metadata file

## Gather file paths 

```{r}
munged_files <- list_sumstats(save_dir = save_dir)
```

## Download OpenGWAS metadata

```{r}
metagwas_all <- MungeSumstats::find_sumstats(ids = names(munged_files))
metagwas_all$munged_path <- munged_files
```

## Get genome builds

### Gather from log files

```{r}
log_data <- parse_logs(save_dir = save_dir)
knitr::kable(log_data)
```

### Infer from data

`MungeSumstats` automatically keeps log files in the `tempdir()`. 
But if you restart your R session mid-way, they'll be erased. 
Or perhaps you ran `MungeSumstats` without turning on any of the log arguments.

Instead, you can infer the genome build directly from the data. 
Note that this will take some time as the genome reference data needed 
to make these inferences is quite large.

```R
builds <- MungeSumstats::get_genome_builds(sumstats_list = munged_files, 
                                           header_only = 1000, 
                                           names_from_paths =TRUE)
table(metagwas_all$build_inferred)
```

## Merge metadata

Now we will merged the metadata from OpenGWAS and 
our parsed `MungeSumstats` logs. 

We will use this information to create a new column that examine
whether the genome-build `MungeSumstats` inferred from the each GWAS dataset (*"build_inferred"*) matches the build reported by OpenGWAS (*"build"*).

```{r}
meta <- merge(metagwas_all, log_data, by = "id")
#### Find which GWAS have different inferred genome builds ####
meta$build_matches <- toupper(
  stringr::str_split(meta$build,"/",
                     simplify = TRUE)[,2]
  )==meta$build_inferred
#### Summarise ####
message(sum(meta$build_matches)," / ",nrow(meta),
        " inferred genome builds match the reported genome build.")
```


## Save metadata
 
```{r} 
save_meta <- file.path(save_dir,"metadata.csv")
utils::write.csv(meta,
                 save_meta)
```


# Map SNPs to genes

## Read metadata

```{r}
# meta <- read.csv(save_meta)
meta$trait_group <- stringr::str_extract(
  tolower(meta$trait),
  "(alzheimer)|(parkinson)|(amyotrophic)|(frontotemporal dementia)|(dementia)")
meta[is.na(meta$trait_group),"trait_group"] <- "other"

```

 
## Map SNPs

```{r}
gene_files <- list_snps_to_genes_files(save_dir = save_dir)

magma_files <- parallel::mclapply(seq(1,nrow(meta)), function(i){
  MAGMA.Celltyping:::message_parallel("----- ", meta$id[i]," -----")
  if(meta$id[i] %in% names(gene_files)) {
     MAGMA.Celltyping:::message_parallel("Importing existing annot file.")
    return(gene_files[[meta$id[i]]])
  } else{
     tryCatch(expr = {
       MAGMA.Celltyping::map_snps_to_genes(path_formatted = meta$munged_path[i],
                                      genome_build = meta$build_inferred[i],
                                      N = meta$N[i],
                                      upstream_kb = 35,  
                                      downstream_kb = 10, 
                                      force_new = FALSE)
     }, error = function(e) NULL)
  }
}) %>% `names<-`(meta$id)
```

# Gather snps2genes files

Now that all the GWAS have been mapped from SNPs to genes, 
we can copy the smaller mapping files into a folder that will 
get pushed to GitHub (i.e. isn't included in the *.gitignore* file).

```{r}
gene_files <- list_snps2genes_files(save_dir = save_dir)
gene_files2 <- copy_snps_to_genes_files(gene_files = gene_files, 
                                        save_dir = here::here("data/MAGMA_Files"), 
                                        overwrite = TRUE)

```



# Construct gene x trait matrix

Can construct gene x trait matrix from gene-level MAGMA results 
across multiple GWAS traits.

```R
magma_matrix <- phenomix::gene_trait_matrix(magma_files = magma_files,
                                            metric = "ADJ_ZSTAT", 
                                            save_path = file.path(save_dir,"OpenGWAS_neurodegenGWAS_ADJZSTAT.rds"),
                                            mc.cores = 10) 
```


# Session Info

<details>

```{r}
utils::sessionInfo()
```

</details>
