---
title: "Manual GWAS"
author: "<h4>Author: <i>Brian M. Schilder</i></h4>"
date: "<h4>Updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output:
  BiocStyle::html_document:
vignette: >
    %\VignetteIndexEntry{OpenGWAS} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)

save_dir <- here::here("data","GWAS_munged")
source(here::here("code","utils.R"))
```

## Install deps 

```{r}
if(!require("MungeSumstats")) remotes::install_github("neurogenomics/MungeSumstats",dependencies = TRUE)
if(!require("downloadR")) remotes::install_github("RajLabMSSM/downloadR")
```

# Pardinas2018

- [Publication](https://doi.org/10.1038/s41588-018-0059-2)
- [Data](https://walters.psycm.cf.ac.uk/)

```{r}
source(here::here("code","prepare_Pardinas2018.R"))
Pardinas2018 <- prepare_Pardinas2018()
```


# Yengo2018

- [Publication](https://doi.org/10.1093/hmg/ddy271)
- [Data](https://portals.broadinstitute.org/collaboration/giant/index.php/GIANT_consortium_data_files)

```{r}
source(here::here("code","prepare_Yengo2018.R"))
Yengo2018 <- prepare_Yengo2018()
```

# Wheeler2017

- [Publication](https://doi.org/10.1371/journal.pmed.1002383)
- [Data](http://magicinvestigators.org/downloads/)

```{r}
source(here::here("code","prepare_Wheeler2017.R"))
Wheeler2017 <- prepare_Wheeler2017()
```


# Nelson2017

- [Publication](https://doi.org/10.1038/ng.3913)
- [Data](http://www.cardiogramplusc4d.org/data-downloads/)

```{r}
source(here::here("code","prepare_Nelson2017.R"))
Nelson2017 <- prepare_Nelson2017()
```

# Jansen2019

- [Publication](https://doi.org/10.1038/s41588-018-0311-9)
- [Data](https://ctg.cncr.nl/software/summary_statistics)

```{r}
source(here::here("code","prepare_Jansen2019.R"))
Jansen2019 <- prepare_Jansen2019()
```

# Wightman2021

- [Publication](https://doi.org/10.1038/s41588-021-00921-z)
- [Data](https://ctg.cncr.nl/software/summary_statistics)

```{r}
source(here::here("code","prepare_Wightman2021.R"))
Wightman2021 <- prepare_Wightman2021()
```
 
# Vuckovic2020

Generate weighted gene lists for each blood trait GWAS from:  
> Vuckovic D, et al. The Polygenic and Monogenic Basis of Blood Traits and Diseases. Cell. 2020 Sep 3;182(5):1214-1231.e11. [doi: 10.1016/j.cell.2020.08.008](https://doi.org/10.1016/j.cell.2020.08.008)

Data source: 
- [UK Biobank subset](ftp://ftp.sanger.ac.uk/pub/project/humgen/summary_statistics/UKBB_blood_cell_traits/): 
- [Meta-analyses (link broken)](http://www.mhi-humangenetics.org/en/resources): 746,667 subjects across 5 ancestries.
- [Trait metadata key](https://www.cell.com/cms/10.1016/j.cell.2020.08.008/attachment/ab993af6-e2f4-4681-9308-30aaa50d3953/mmc1.xlsx)

Below, we analyse only the UK Biobank subset of the results, as the link to the full meta-analyses are now broken. 

```{r}
source(here::here("code","prepare_Vuckovic2020.R"))
Vuckovic2020 <- prepare_Vuckovic2020()
```

# Copy snps2genes files

Now that all the GWAS have been mapped from SNPs to genes, 
we can copy the smaller mapping files into a folder that will 
get pushed to GitHub (i.e. isn't included in the *.gitignore* file).

```{r}
gene_files2 <- copy_snps_to_genes_files(search_dir = save_dir,
                                        save_dir = here::here("data","MAGMA_Files"), 
                                        overwrite = FALSE) 
```

<hr> 

# Session Info

<details>

```{r}
utils::sessionInfo()
```

</details>

<br>

