---
title: "OpenGWAS"
author: "<h4>Author: <i>Brian M. Schilder</i></h4>"
date: "<h4>Updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output:
  BiocStyle::html_document:
vignette: >
    %\VignetteIndexEntry{OpenGWAS} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---

# Install packages

To install the dev branch of `MAGMA.Celltyping`, use:

```{r, eval=FALSE, include=TRUE}
if(!require("remotes")) install.packages("remotes")
if(!require("here")) install.packages("here")
if(!require("BiocManager")) install.packages("BiocManager")

if(!require("MAGMA.Celltyping")) {
  remotes::install_github("neurogenomics/MAGMA_Celltyping", dependencies = TRUE)
}
if(!require("MungeSumstats") | packageVersion("MungeSumstats")<"1.3.17") {
   remotes::install_github("neurogenomics/MungeSumstats", dependencies = TRUE)
}
if(!require("EWCE") | packageVersion("EWCE")<"1.3.1"){
  remotes::install_github("NathanSkene/EWCE", dependencies = TRUE)
} 
```


```{r setup, include=TRUE}
library(MAGMA.Celltyping)
library(MungeSumstats)
library(data.table)
library(ggplot2)
library(dplyr)

source(here::here("code/utils.R"))

save_dir <- here::here("data","GWAS_munged")
dir.create(save_dir, showWarnings = FALSE, recursive = TRUE)
```

# Munge sum stats

## Find sum stats

### Neurodegeneration traits

First, we will search for GWAS of several neurodegenerative disease traits.

```{r} 
traits1 <- c("Alzheimer","Parkinson","amyotrophic",
             "frontotemporal dementia","dementia")
metagwas1 <- MungeSumstats::find_sumstats(traits = traits1)
```

### BMI traits

Next, we perform a second search for Body Mass Index (BMI)-related traits.
We are querying these separately because we want to apply some additional filtering such that we only include GWAS with the greatest sample sizes from the UK Biobank, as well as exclude GWAS of "weight *loss*" or "weight *change*". 

```{r}
traits2 <- c("body mass index","standing height","weight")
metagwas2 <- MungeSumstats::find_sumstats(traits = traits2)
metagwas2 <- filter_traits(meta = metagwas2, 
                            traits = traits2, 
                            consortia = c("Neale Lab","UK Biobank"),
                            exclusion_terms = c("loss","change"),
                            startswith_only = TRUE,
                            topn = 1) 
```

### Sampling of neuro/immuno/cardio traits

Take only the top 100 most well-powered studies, to start.

```{r}
subcategories3 <- c("neurological","Immune","cardio")
metagwas3 <- MungeSumstats::find_sumstats(subcategories = subcategories3)
metagwas3 <- filter_traits(meta = metagwas3,  
                           group_var = "subcategory",
                           topn = 100) 
```

## All UKB traits

```{r}
metagwas4  <- MungeSumstats::find_sumstats(consortia = c("Neale Lab","UK Biobank"))
```

Finally, we can merge the metadata from our two GWAS searches.
We use th `assign_trait_groups` functions to automatically assign mutually-exclusive groups to each of the traits based on our search terms. 
**NOTE**: You may want to double check the "trait_group" assignments, as these may not always be what you had in mind; e.g. "frontotemporal dementia (FTD)" might get assigned to the generic "dementia" group instead of the more appropriate "frontotemporal dementia" group simply due to the order in which groups were assigned.

```{r}
# metagwas5 <- MungeSumstats::find_sumstats(consortia = "PGC")

metagwas <- rbind(metagwas1,metagwas2,metagwas3,metagwas4, fill=TRUE)
traits <- c(traits1,traits2,subcategories3,"UKB")
meta <- assign_trait_groups(meta = metagwas, 
                            traits = traits)
knitr::kable(meta)
```

```{r}
gd <- googledrive::drive_download("https://docs.google.com/spreadsheets/d/1zPhp1daQAs4JN88t4VuLbhpc8ygKFOsA-IdVWQa8LRc/edit#gid=1126020500")
meta <- readxl::read_excel(gd$local_path,sheet = "GWAS")
meta <- data.table::fread(here::here("code/gwas_meta.tsv"))
```


## Import sum stats

`MungeSumstats::import_sumstats` already supports parallelisation 
across multiple GWAS (by setting `nThread` >1). 
However, the log files would all 
overwrite one another. 

So we instead parellelise this step ourselves,
saving each summary stats file in it's own folder with a 
subdirectory for its respective log files.

## OpenGWAS

### Errors

#### Error 1
- `BiocParallel errors
  0 remote errors, element index: 
  50 unevaluated and other errors
  first remote error:`
- No sumstats produced.
- Also recorded messages from the wrong/next id!

#### Error 2
- ukb-a-367
-  `[E::bgzf_read_block] Failed to read BGZF block data at offset 247839653 expected 17773 bytes; hread returned 683`
 
   
```{r}
# munged_files <- list_sumstats(save_dir = save_dir)
#### First, let's remove incomplete files.  ####
# bad_files <- munged_files[file.info(munged_files)$size <100]
# file.remove(bad_files)

# unlink("data/GWAS_munged/",recursive = TRUE, force = TRUE); 
# unlink("data/VCFs/",recursive = TRUE, force = TRUE); 

gwas_paths <- MungeSumstats::import_sumstats(
  ids = meta$id, 
  save_dir = here::here("data/GWAS_munged"), 
  nThread = 8, # >30 causes issues with read_vcf_parallel
  parallel_across_ids = FALSE,
  dbSNP = 144,
  force_new_vcf = FALSE,
  force_new = FALSE,
  vcf_download = TRUE,
  vcf_dir = here::here("data/VCFs"),
  ### axel will keep trying forever if the URL doesn't exist (or is private)
  # download_method = "axel",
  #### Record logs
  log_folder_ind = TRUE,
  log_mungesumstats_msgs = TRUE,
  ) 
```


# Create metadata file 

```{r}
meta <- gather_metadata(save_dir = save_dir,
                        N_dict=c("Wightman2021"=1126563, 
                                 "Vuckovic2020"=408112)) 
```

# Map SNPs to genes  

Parallelising `map_snps_to_genes` across too many threads actually seems to slow everything down, despite having sufficient memory.

This might be because:

1. They are all calling to the same underlying MAGMA C code.
2. **They are reserving more memory in C than the computer thinks is available.** (seems likely)
3. They are all reading from the same reference files (e.g. 1KG).
4. MAGMA is somehow already parallelising processes internally. 
5. Docker is launching some processes that don't get killed when restarting the container.
6. Intermediate MAGMA files are getting repeatedly appended and causing problems in subsequent reruns.
7. `genes_only=FALSE`? but this should make things faster, not slower...Indeed, setting `genes_only=TRUE` seems to increase time.
8. Newer versions of MAGMA(v1.10) has bugs that causes everything to slow down.
  - MAGMAv1.08 might be a bit faster, but hard to tell bc uses diff files each time. 
  
**Solution**:
 
 Limit the number of parallel threads to <=20. This results in low memory usage (~11/252GB at any given time, indicated by the green bar in `htop`) but, perhaps importantly, ensures that the amount of memory being reserved only reaches ~2/3 of the max (the yellow bar in `htop`).

```{r}
# gene_files <- list_snps_to_genes_files(save_dir = save_dir)
# MAGMA.Celltyping::magma_install(desired_version = "1.08", upgrade = TRUE)

t1 <- Sys.time()
magma_files <- parallel::mclapply(seq_len(nrow(meta)),
                                  function(i){ 
  EWCE:::message_parallel("----- ",i," : ",
                                      meta$id[i]," -----") 
     tryCatch(expr = { 
       MAGMA.Celltyping::map_snps_to_genes(
         # version = "1.08",
         path_formatted = meta$munged_path[i],
         genome_build = meta$build_final[i],
         N = if(is.na(meta$N[i])) NULL else meta$N[i],
         population = meta$population_1KG[i],
         upstream_kb = 35,
         downstream_kb = 10,
         genes_only = TRUE, 
         force_new = FALSE
         )
     }, error = function(e) {EWCE:::message_parallel(e);NULL}) 
}, mc.cores = min(nrow(meta),20) ) |> `names<-`(meta$id) 
t2 <- Sys.time()
print(t2-t1)

```

# Gather snps2genes files

Now that all the GWAS have been mapped from SNPs to genes, 
we can copy the smaller mapping files into a folder that will 
get pushed to GitHub (i.e. isn't included in the *.gitignore* file).

```{r}
gene_files2 <- copy_snps_to_genes_files(search_dir = save_dir,
                                        save_dir = here::here("data/MAGMA_Files"), 
                                        overwrite = FALSE)

#### Find smallest files for demo purposes ####
# f <- unlist(gene_files2)
# data.frame(file=f,
#            size=file.info(f)$size) %>% dplyr::arrange(size)
```

## Re-save metadata

Add the paths of the gene mapping files to the metadata and resave it.

We will also add a URL that can be used to download the files. 

```{r}
meta <- gather_metadata(save_dir = save_dir)
```


# Studies over time

```{r}
meta <- MungeSumstats::find_sumstats()

counts <- dplyr::group_by(meta, year) |>
  dplyr::summarise(datasets=dplyr::n_distinct(id),
                   participants=sum(sample_size, na.rm=TRUE)) |>
  dplyr::filter(!is.na(year)) 

gp_gwas <- ggplot(counts, aes(x=year,
                   y=cumsum(datasets), 
                   fill=cumsum(participants))
       ) +
  geom_col() +  
  geom_point() +
  geom_line(alpha=.5) + 
  scale_x_continuous(breaks=seq(min(counts$year),
                              max(counts$year), 4), 
                     limits = c(2002,NA)) + 
  scale_fill_viridis_c(option="plasma") +
  theme_bw() +
  labs(title="GWAS")
```

## Single-cell RNA-seq data

```{r}
sc <- data.table::fread("http://www.nxn.se/single-cell-studies/data.tsv")
sc$year <- as.numeric(substr(sc$Date,1,4))

sc_counts <- dplyr::group_by(sc, year) |>
  dplyr::summarise(studies=dplyr::n_distinct(Shorthand), 
                   celltypes=sum(as.numeric(`Number of reported cell types or clusters`), na.rm = TRUE))


gp_sc <- ggplot(sc_counts, aes(x=year,
                   y=cumsum(studies),
                   fill=cumsum(celltypes)
               )
       ) +
  geom_col() +  
  geom_point() +
  geom_line(alpha=.5) + 
  scale_x_continuous(breaks=seq(min(counts$year),
                              max(counts$year), 4)) + 
  scale_fill_viridis_c(option="plasma") +
  theme_bw() +
  labs(title="scRNA-seq")

```

 
```{r}
plts <- list("GWAS"=gp_gwas, "scRNA-seq"=gp_sc)
patchwork::wrap_plots(plts,
                      ncol = 1) 
```
 

# Session Info

<details>

```{r}
utils::sessionInfo()
```

</details>

<hr>
