---
title: "metadata_report"
author: "<h4>Authors: <i>Brian M. Schilder, Alan Murphy</i></h4>"
date: "<h4>Updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output: rmarkdown::html_document
vignette: >
    %\VignetteIndexEntry{metadata_report} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
--- 

```{r setup, include=TRUE}
library(here)
library(ggplot2)
library(dplyr)
library(data.table)
```

# Background

Analysis of metadata from [MungeSumstats](https://academic.oup.com/bioinformatics/article/37/23/4593/6380562)
quality control and standardisation of 714 GWAS summary statistics from IEU GWAS.


# Import metadata 

```{r}
meta <- data.table::fread(here::here("metadata.csv"), drop = "V1")

knitr::kable(head(meta))
```



# Plot 

## Categorical metadata

First, let's get a sense of the characteristics of the different GWAS summary 
statistics:

```{r, fig.height=10, fig.width=14}
variables <- c("category", "subcategory", "consortium", "population", "unit")
# palettes <- rownames(subset(RColorBrewer::brewer.pal.info, category=="div"))
palettes <- c(pals::stevens.pinkblue, 
              pals::arc.bluepink,
              pals::brewer.bugn,
              pals::brewer.seqseq2,
              pals::brewer.bupu)

gg_pies <- lapply(seq_len(length(variables)), function(i){
  v <- variables[i]
  pal <- palettes[i]
  n_colors <- dplyr::n_distinct(meta[[v]])
  message(v)
  #### Prepare data ####
  pie_data <- meta %>% 
  dplyr::mutate_at(.vars = dplyr::all_of(v), 
                   .funs = function(v){ifelse(v=="NA", NA, v)}) %>%
  dplyr::group_by_at(dplyr::all_of(v)) %>%
  dplyr::count() %>%
  data.table::data.table() %>% 
  dplyr::mutate_at(.vars = v,
                   .funs = stringr::str_trunc, 25)
  #### Plot ####                  
  ggplot(pie_data, 
         aes_string(x=1, y="n", fill=v)) + 
  geom_bar(width = 1, stat = "identity") +
  coord_polar(theta = "y") + 
  labs(title=v, x=NULL, y=NULL) + 
  theme_bw()+ 
  theme(aspect.ratio = 1, 
        legend.position = "bottom",
        legend.title = element_blank(), 
        axis.text.x = element_blank(),
        axis.text.y = element_blank()) +
  guides(fill = guide_legend(ncol = 2)) + 
  scale_fill_manual(values = palettes[[i]](n = n_colors))
}) %>% `names<-`(variables)

patchwork::wrap_plots(gg_pies, nrow = 1)

```

## Over time

```{r, fig.height=10, fig.width=10}
variables <- c("nsnp","N","population")
time_data <- meta %>% 
  dplyr::select(id, year, all_of(variables)) %>%
  dplyr::mutate(year=factor(year,
                            levels = sort(unique(meta$year)),
                            ordered = TRUE), 
                population = stringr::str_trunc(population, 25))

gg_time <- lapply(seq_len(length(variables)), function(i){
  v <- variables[i]
  message(v)
  ggplot(time_data, 
         aes_string(x="year", y=v, color=v)) +
    geom_boxplot() + 
    geom_point(alpha=.5) +
    geom_jitter(height = 0) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(title=v)
}) %>% `names<-`(variables)

patchwork::wrap_plots(gg_time, ncol = 1)

```

## Genome builds

All genome builds on IEU GWAS should be hg19/GRCh37, is this actually the case?:

```{r}
ggplot(meta,aes(x=build_inferred,fill=build_inferred))+
  geom_bar()+
  geom_text(aes( label = paste0(format(100*..count../nrow(meta), digits=2, 
                                  drop0trailing=TRUE),"% - ",..count..),
                 y= ..count.. ), stat= "count", vjust = -.5)+
  scale_fill_manual(values=c("purple","orange"))+
  theme_bw()+
  labs(y= "Number of Studies", 
       x = "Inferred Genomic Build") +
  theme(strip.background=element_blank(),
        legend.position = "none")
```

So 25 studies of what we evaluated actually are GRCh38!

More detail:

```{r, fig.height=10, fig.width=10}
library(ggalluvial) 
alluv_data <- meta %>% 
  dplyr::select(build, build_inferred, build_matches, category) %>%
  dplyr::mutate(n=1, 
                build = gsub("HG19/GRCh37","GRCH37",build))

# ggalluvial::is_alluvia_form(alluv_data)
ggplot(alluv_data, 
       aes(y = n, 
           axis1 = category,
           axis2 = build_inferred)) +
  ggalluvial::geom_alluvium(aes(fill=build_matches)) +
  ggalluvial::geom_stratum(width = 1/12, fill = "black", color = "grey") +
  geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
  scale_x_discrete(limits = c("build", "build_inferred"), expand = c(.05, .05)) +
  scale_fill_brewer(type = "qual", palette = "Set1") +
  theme_bw() +
  labs(title="Genome builds: OpenGWAS metadata vs. Inferred")
```


## SNP filtering 

```{r}
variables <- sort(grep("snps",colnames(meta), value = TRUE),
                  decreasing = TRUE)
filt_data <- meta[,..variables]
```

### Quality Control of SNPs

Overview charts:

```{r, fig.height=10, fig.width=10}
gg_hist <- lapply(seq_len(length(variables)), function(i){
  v <- variables[i]
  message(v)
  ggplot(filt_data, 
         aes_string(x = v)) +
  geom_histogram(bins = 75) + 
  theme_bw() 
}) %>% `names<-`(variables)

patchwork::wrap_plots(gg_hist, ncol = 2)
```

How many SNPs fail quality control checks?:

```{r}
meta[,snps_dropped_total:=nsnp-snps_end]
#add constand column for plot
meta[,sumstats:="sumstats"]
ggplot(meta,aes(x=sumstats,y=snps_dropped_total/nsnp,fill=sumstats))+
  geom_violin()+
  geom_boxplot(width=0.05)+
  theme_bw()+
  labs(y= "Proportion of SNPs failed quality control", x = "") +
  theme(strip.background=element_blank(),
        legend.position = "none")
```


And why are these SNPs removed?:

```{r}
snps_dropped <- melt(meta[,c("id","nsnp","snps_dropped_nonA1A2",
                             "snps_dropped_nonRef","snps_dropped_chrom",
                             "snps_dropped_INFO","snps_dropped_duplicates")],
                     id.vars=c("id","nsnp"))
#remove snps_dropped_
snps_dropped[,variable:=substring(variable, 14)]

ggplot(snps_dropped,aes(x=variable,y=value/nsnp,fill=variable))+
  geom_violin()+
  geom_boxplot(width=0.02)+
  theme_bw()+
  labs(y= "Proportion of SNPs failed quality control", 
       x = "Exclusion criteria") +
  theme(strip.background=element_blank(),
        legend.position = "none")

```

INFO filtering is the biggest determinant, if we remove and replot?

```{r}

ggplot(snps_dropped[variable!="INFO"],
        aes(x=variable,y=value/nsnp,fill=variable))+
  geom_violin()+
  geom_boxplot(width=0.02)+
  theme_bw()+
  labs(y= "Proportion of SNPs failed quality control", 
       x = "Exclusion criteria") +
  theme(strip.background=element_blank(),
        legend.position = "none")

```


Further detail:

```{r, fig.height=10, fig.width=10}
filt_data2 <- filt_data %>% data.table::melt.data.table()

gg_filt <- ggplot(filt_data2, aes(x=value, y=variable)) +
  geom_boxplot() +
  geom_point(alpha=.5) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle=45, hjust=1))

gg_filt
```


## Effect direction of SNPs

The effect direction or SNPs is also evaluated against a reference genome and in
cases where it is incorrect, it is corrected. Let's check how many are 
corrected:

```{r}

ggplot(meta,aes(x=snps_flipped/nsnp,fill=sumstats))+
  geom_histogram(bins = 75)+
  theme_bw()+
  labs(x= "Proportion of SNPs flipped", x = "Number of studies") +
  theme(strip.background=element_blank(),
        legend.position = "none")

```


There are about 25/30 GWAS sumstats which contain SNPs with an incorrect 
direction (accounting for about 18% of all SNPs in those studies).

# Session Info

<details>

```{r}
utils::sessionInfo()
```

</details>

<br>
